app/scenes/login/login.js

```diff
@@ -39,11 +58,69 @@ class Login extends Component {
         }
     }
 
-    signIn() {
+    preSignIn = () => {
+        this.setState({error: null});
+        if (!this.props.loginId) {
+            // it's slightly weird to be constructing the message ID, but it's a bit nicer than triply nested if statements
+            let msgId = 'login.no';
+            if (this.props.config.EnableSignInWithEmail === 'true') {
+                msgId += 'Email';
+            }
+            if (this.props.config.EnableSignInWithUsername === 'true') {
+                msgId += 'Username';
+            }
+            if (this.props.license.IsLicensed === 'true' && this.props.config.EnableLdap === 'true') {
+                msgId += 'LdapUsername';
+            }
+
+            this.setState({
+                error: (
+                    <FormattedText
```

Comment:
Seems odd to be storing an entire component in the state. Is there not a way to just store the id and default message in the state instead of the entire component?